src/Agda2Hs/Compile/RuntimeCheckUtils.hs

```diff
@@ -0,0 +1,108 @@
+import Agda.Syntax.Internal
+import Agda.TypeChecking.InstanceArguments (findInstance)
+import Agda.TypeChecking.MetaVars (newInstanceMeta, newLevelMeta)
+import Agda.TypeChecking.Monad (addContext, liftTCM)
+import Agda.TypeChecking.Pretty (prettyTCM)
+import Agda.TypeChecking.Reduce (instantiate)
+import Agda.TypeChecking.Telescope (splitTelescopeAt)
+import Agda2Hs.Compile.Term (compileTerm)
+import Agda2Hs.Compile.Types (C)
+import Agda2Hs.Compile.Utils
+import Agda2Hs.HsUtils
+import Control.Monad.Except (catchError)
+import Data.List (partition)
+import Data.Maybe (catMaybes, isJust)
+import qualified Language.Haskell.Exts as Hs
+
+-- Checks if an argument is erased
+checkErased :: Dom (ArgName, a) -> Bool
+checkErased t = case getQuantity t of
+  Quantity0 _ -> True
+  _ -> False
+
+-- Convenience function to prepend signatures if the check exists
+maybeAppend :: [a] -> Maybe a -> Maybe [a]
+maybeAppend sig chk = (sig ++) . (: []) <$> chk
+
+-- Replace unnamed arguments with a${i}
+-- NB this does not handle nested but neither does the encapsulating logic yet
+nameArgs :: [(Dom (ArgName, a), Int)] -> [String]
+nameArgs = map $ \(d, i) -> let name = fst $ unDom d in if name == "_" then "a" ++ show i else name
+
+-- Turn a type into its Dec version
+decify :: Type -> C Type
+decify t = do
+  dec <- resolveStringName "Haskell.Extra.Dec.Dec"
+  level <- liftTCM newLevelMeta
+  let vArg = defaultArg
+      hArg = setHiding Hidden . vArg
+  return $ t {unEl = Def dec $ map Apply [hArg $ Level level, vArg $ unEl t]}
+
+-- Failably find instances for decidable terms
+findDecInstances :: Type -> C (Maybe Term)
+findDecInstances t =
+  liftTCM $
+    do
+      (m, v) <- newInstanceMeta "" t
+      findInstance m Nothing
+      Just <$> instantiate v
+      `catchError` (\_ -> return Nothing)
+
+-- Create expression to be put in the guard
+createGuardExp :: Type -> C (Maybe (Hs.Exp ()))
+createGuardExp t = do
+  dec <- decify t
+  sth <- findDecInstances dec
+  -- XXX functor something?
+  findDecInstances dec >>= \case
+    Just r -> Just <$> compileTerm dec r
+    Nothing -> return Nothing
+
+-- Create run time check.
+-- Takes function name, arguments, expression on success, and expressions for checks.
+createRtc :: Hs.Name () -> [String] -> Hs.Exp () -> [Hs.Exp ()] -> Hs.Decl ()
+createRtc n args success chks =
+  Hs.FunBind () [Hs.Match () n (map (Hs.PVar () . hsName) args) rhs Nothing]
+  where
+    andChks = foldr1 (\c -> Hs.InfixApp () c $ Hs.QVarOp () $ Hs.UnQual () $ Hs.Symbol () "&&") chks
+    errChks =
+      map
+        ( \chk ->
+            let msg = "Runtime check failed: " ++ Hs.prettyPrint chk
+                hsErr = Hs.App () (hsVar "error") $ Hs.Lit () $ Hs.String () msg msg
+             in Hs.GuardedRhs () [Hs.Qualifier () $ Hs.App () (hsVar "not") chk] hsErr
+        )
+        chks
+    successApp = eApp success $ map hsVar args
+    rhs = Hs.GuardedRhss () (Hs.GuardedRhs () [Hs.Qualifier () andChks] successApp : errChks)
```

Comment:
This is all a bit wordy with the Haskell constructs but I suppose there isn't really a way around it.